### PR TITLE
Consolidate contiguous var declarations in destructuring transform

### LINKED ISSUE #3081: Regression: Variable definitions in a For initialiser loop are put inside an IIFE
With the latest version of babel and its presets (`6.1.18 (babel-core 6.1.21)`), running `babel --preset es2015 script.js` with `script.js` containing:
```js
let list = [1, 2, 3, 4]
for ( let i = 0, { length } = list; i < length; i++ ) {
	console.log( i + ': ' + list[i] )
}
```
…gives:
```js
'use strict';

var list = [1, 2, 3, 4];
for ((function () {
	var i = 0;var length = list.length;
})(); i < length; i++) {
	console.log(i + ': ' + list[i]);
}
```
…which is totally wrong. The IIFE hides the declared variables from the for-loop body.

This issue wasn't there in version `6.0.14` at least.

### PR DESCRIPTION
**Pull request by motiz88**
| Q                 | A
| ----------------- | ---
| Bug fix?          | yes
| Breaking change?  | no
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | yes
| Fixed tickets     | #3081 
| License           | MIT

Fixes #3081 by never trying to inject multiple VariableDeclaration nodes into a `for` loop's init clause - we now always deduplicate them to a single node because they will all have the same `kind`. The order is maintained so execution order of any init expressions is the same.

This solution has the stylistic side effect of reducing the number of `var`s in the output, and neatly mapping each input declaration to one grouped output declaration.

I purposely didn't touch the core destructuring logic in any way and implemented this as a second pass over the generated `nodes` array. If anyone prefers a different implementation, let me know!

### CODE REVIEW

**Comment by motiz88** [comment]
I don't think that would ever be hit. Actually, I think the `nodes.length !== 1` branch is dead code now, which means the new loop can almost definitely be folded into the one above it somehow and `nodes` replaced with a scalar; I'm just not sure how the upper loop works exactly (what does `t.inherits` do? etc)

**Comment by hzoo** [comment]
I'm not sure yet either - we can do a fix in another PR as followup though?

File: packages/babel-plugin-transform-es2015-destructuring/src/index.js [hunk 1]
```diff
@@ -492,7 +492,21 @@ export default function ({ types: t }) {
           }
         }
 
-        path.replaceWithMultiple(nodes);
+        const nodesOut = [];
+        for (const node of nodes) {
+          const tail = nodesOut[nodesOut.length - 1];
+          if (tail && tail.kind === node.kind) {
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-destructuring/src/index.js]: Right I think that's what @motiz88 was getting at in the comment ^

**Comment by motiz88** [comment]
On it.

**Comment by motiz88** [comment]
Okay, seems like my assumption that `nodes` are all VariableDeclarations of the same `kind` [here](https://github.com/motiz88/babel/blob/15cb373f0726f68225f7080a7ae206a63af174ee/packages/babel-plugin-transform-es2015-destructuring/src/index.js#L495) is false in a number of cases (see [test failures](https://travis-ci.org/babel/babel/jobs/167389463) for 15cb373). In some places [`kind` has different values](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-es2015-modules-commonjs/test/fixtures/regression/T7199/expected.js#L11); worse, in others we have [embedded assignment statements](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-es2015-destructuring/test/fixtures/destructuring/assignment-expression-pattern/expected.js#L3).

So at this point, I think optimizing beyond what we had in 94a6b23 would require some serious (painful) digging around in `DestructuringTransformer` for no clear benefit. I'm gonna go ahead and revert 15cb373, and then I think we're good to merge.

**Pull request merged**